drivers/sensor/tdk/icp101xx/icp101xx.c

```diff
@@ -0,0 +1,224 @@
+			return -EINVAL;
+		}
+	};
+	return err;
+}
+
+static int icp101xx_sample_fetch(const struct device *dev, const enum sensor_channel chan)
+{
+	icp101xx_data *data = (icp101xx_data *)dev->data;
+	int rc = 0;
+	uint64_t timeout;
+
+	if (!(chan == SENSOR_CHAN_AMBIENT_TEMP || chan == SENSOR_CHAN_PRESS ||
+	      SENSOR_CHAN_ALTITUDE || chan == SENSOR_CHAN_ALL)) {
+		return -ENOTSUP;
+	}
+	rc = inv_icp101xx_enable_sensor(&data->icp_device, 1);
+	/* Compute timeout for the measure */
+	timeout = k_uptime_get() + get_timeout_ms(data->icp_device.measurement_mode);
+	/* Initial sleep waiting the sensor proceeds with the measure */
+	k_sleep(K_MSEC(get_conversion_ms(data->icp_device.measurement_mode)));
+	do {
+		k_sleep(K_USEC(200));
+		rc = inv_icp101xx_get_data(&data->icp_device, &(data->raw_pressure),
+					   &(data->raw_temperature), &(data->pressure),
+					   &(data->temperature));
+	} while ((rc != 0) && (k_uptime_get() <= timeout));
+	return rc;
+}
+
+static int icp101xx_channel_get(const struct device *dev, enum sensor_channel chan,
+				struct sensor_value *val)
+{
+	float pressure;
+	icp101xx_data *data = (icp101xx_data *)dev->data;
+
+	if (!(chan == SENSOR_CHAN_AMBIENT_TEMP || chan == SENSOR_CHAN_PRESS ||
+	      SENSOR_CHAN_ALTITUDE)) {
+		return -ENOTSUP;
+	}
+	/* Zephyr expects kPa while ICP101xx returns Pa */
+	pressure = data->pressure / 1000;
+	if (chan == SENSOR_CHAN_AMBIENT_TEMP) {
+		sensor_value_from_float(val, data->temperature);
+	} else if (chan == SENSOR_CHAN_PRESS) {
+		sensor_value_from_float(val, pressure);
+	} else if (chan == SENSOR_CHAN_ALTITUDE) {
+		float altitude = convertToHeight(pressure, data->temperature);
+
+		sensor_value_from_float(val, altitude);
+	} else {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int icp101xx_init(const struct device *dev)
+{
+	struct inv_icp101xx_serif icp101xx_serif;
+	int rc = 0;
+	icp101xx_data *data = (icp101xx_data *)dev->data;
+	const struct icp101xx_config *cfg = (const struct icp101xx_config *)dev->config;
+
+	icp101xx_serif.context = (void *)dev;
+	icp101xx_serif.read_reg = inv_io_hal_read_reg;
+	icp101xx_serif.write_reg = inv_io_hal_write_reg;
+	icp101xx_serif.max_read = 2048;  /* maximum number of bytes allowed per serial read */
+	icp101xx_serif.max_write = 2048; /* maximum number of bytes allowed per serial write */
+	icp101xx_serif.is_spi = 0;
+	/*
+	 * Reset pressure sensor driver states
+	 */
+	inv_icp101xx_reset_states(&(data->icp_device), &icp101xx_serif);
+
+	rc = inv_icp101xx_soft_reset(&data->icp_device);
+	if (rc != 0) {
+		LOG_ERR("Soft reset error %d", rc);
+		return rc;
+	}
+	inv_icp101xx_init(&data->icp_device);
+	if (rc != 0) {
+		LOG_ERR("Init error %d", rc);
+		return rc;
+	}
+	data->icp_device.measurement_mode = cfg->mode;
+
+	/* successful init, return 0 */
+	return 0;
+}
+
+static const struct sensor_driver_api icp101xx_api_funcs = {
+	.sample_fetch = icp101xx_sample_fetch,
+	.channel_get = icp101xx_channel_get,
+	.attr_set = icp101xx_attr_set,
```

Comment:
I'd really recommend implementing the submit and decoder APIs as well, as those are the future of Zephyr sensor drivers. In which case for this device on a request to read you would...

* Enable the device to sample
* Start a k_timer
* In the k_timer callback start a i2c read to directly read into a user given buffer
* On i2c read complete you'd mark the read complete

In the decoder you'd implement the ability to then decode the sensor data from its internal register format, stashed in the buffer, to useful fixed point q values.